From 72383bc27a0dc5646644a7b4166247049ab09416 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:38:44 -0400
Subject: [virt] kvm/nvmx: Add tracepoints for nested_vmexit and nested_vmexit_inject

Message-id: <1406040016-3289-130-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86208
O-Subject: [RHEL7 PATCH v2 129/221] KVM: nVMX: Add tracepoints for nested_vmexit and nested_vmexit_inject
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jan Kiszka <jan.kiszka@siemens.com>

Already used by nested SVM for tracing nested vmexit: kvm_nested_vmexit
marks exits from L2 to L0 while kvm_nested_vmexit_inject marks vmexits
that are reflected to L1.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 542060ea79c861e100411a5a44df747b56a693df)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 15d006f..aa5fe61 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -6697,6 +6697,13 @@ static bool nested_vmx_exit_handled(struct kvm_vcpu *vcpu)
  struct vmcs12 *vmcs12 = get_vmcs12(vcpu);
  u32 exit_reason = vmx->exit_reason;
 
+ trace_kvm_nested_vmexit(kvm_rip_read(vcpu), exit_reason,
+    vmcs_readl(EXIT_QUALIFICATION),
+    vmx->idt_vectoring_info,
+    intr_info,
+    vmcs_read32(VM_EXIT_INTR_ERROR_CODE),
+    KVM_ISA_VMX);
+
  if (vmx->nested.nested_run_pending)
   return 0;
 
@@ -8468,6 +8475,13 @@ static void nested_vmx_vmexit(struct kvm_vcpu *vcpu, u32 exit_reason,
  prepare_vmcs12(vcpu, vmcs12, exit_reason, exit_intr_info,
          exit_qualification);
 
+ trace_kvm_nested_vmexit_inject(vmcs12->vm_exit_reason,
+           vmcs12->exit_qualification,
+           vmcs12->idt_vectoring_info_field,
+           vmcs12->vm_exit_intr_info,
+           vmcs12->vm_exit_intr_error_code,
+           KVM_ISA_VMX);
+
  cpu = get_cpu();
  vmx->loaded_vmcs = &vmx->vmcs01;
  vmx_vcpu_put(vcpu);
-- 
1.7.1